src/fsharp/FSharp.Core/fslib-extra-pervasives.fs

```diff
@@ -70,6 +81,78 @@ module ExtraTopLevelOperators =
+                            member s.GetEnumerator() = ((keys |> Seq.map (fun v -> v.Value)) :> System.Collections.IEnumerable).GetEnumerator() }
+                    
+                member s.Values =
+                    { new IEnumerable<'T> with
+                            member s.GetEnumerator() = d.Values.GetEnumerator()
+                      interface System.Collections.IEnumerable with
+                            member s.GetEnumerator() = (d.Values :> System.Collections.IEnumerable).GetEnumerator() }
+                member s.ContainsKey(k) = d.ContainsKey(RuntimeHelpers.StructBox(k))
+                member s.TryGetValue(k,r) = 
+                    let key = RuntimeHelpers.StructBox(k)
+                    if d.ContainsKey(key) then (r <- d.[key]; true) else false
+          interface IReadOnlyCollection<KeyValuePair<'Key, 'T>> with         
+                member s.Count = c.Count
+#endif
+          interface System.Collections.IDictionary with 
+                member s.IsReadOnly = true
+                member s.IsFixedSize = true
+                member s.IsSynchronized = true
+                member s.Item 
+                    with get x = d.[RuntimeHelpers.StructBox(x :?> 'Key)] :> obj
+                    and  set x v = raise (NotSupportedException(SR.GetString(SR.thisValueCannotBeMutated)))
+                member s.Keys = 
+                    let keys = d.Keys
+                    { new System.Collections.ICollection with
+                          member s.CopyTo(arr,i) = 
+                              let mutable n = 0 
+                              for k in keys do 
+                                  arr.SetValue(k.Value,i+n)
+                                  n <- n + 1
+                          member s.IsSynchronized = true
+                          member s.SyncRoot = s :> obj
+                          member s.Count = keys.Count
+                      interface System.Collections.IEnumerable with
+                            member s.GetEnumerator() = ((keys |> Seq.map (fun v -> v.Value)) :> System.Collections.IEnumerable).GetEnumerator() }
+                    
+                member s.Values = 
+                    { new System.Collections.ICollection with
+                          member s.CopyTo(arr,i) = 
+                              let mutable n = 0 
+                              for k in d.Values do 
+                                  arr.SetValue(k,i+n)
+                                  n <- n + 1
+                          member s.IsSynchronized = true
+                          member s.SyncRoot = s :> obj
+                          member s.Count = d.Values.Count
+                      interface System.Collections.IEnumerable with
+                            member s.GetEnumerator() = (d.Values :> System.Collections.IEnumerable).GetEnumerator() }
+                member s.Count = c.Count
+                member s.SyncRoot = s :> obj
```

Comment:
`SyncRoot` should generally never be a public object (e.g. `this`), but I guess ok in this case since it's never actually used...